contracts/service-registry/src/contract/query.rs

```diff
@@ -41,10 +45,64 @@ pub fn active_verifiers(
         .into_report()?;
 
     if verifiers.len() < service.min_num_verifiers.into() {
-        Err(report!(ContractError::NotEnoughVerifiers))
-    } else {
-        Ok(verifiers)
+        return Err(report!(ContractError::NotEnoughVerifiers));
     }
+
+    Ok(match service.max_num_verifiers {
+        Some(max_verifiers) => select_top_verifiers(verifiers, max_verifiers, env.block.height),
+        _ => verifiers,
+    })
+}
+
+fn select_top_verifiers(
+    mut verifiers: Vec<WeightedVerifier>,
+    max_verifiers: u16,
+    block_height: u64,
+) -> Vec<WeightedVerifier> {
+    if verifiers.len() <= max_verifiers as usize {
+        return verifiers;
+    }
+
+    // Figure out the min weight that will make it into the set
+    let cutoff_weight = {
+        let nth = max_verifiers as usize - 1;
+        verifiers.select_nth_unstable_by(nth, |a, b| b.weight.cmp(&a.weight));
+        verifiers[nth].weight
+    };
+
+    // Now sort by weight descending, but for verifiers with the cutoff weight, break ties via pick_random
+    verifiers.sort_by(|a, b| {
+        b.weight.cmp(&a.weight).then_with(|| {
+            if a.weight == cutoff_weight {
+                pick_random(a, b, block_height)
+            } else {
+                Ordering::Equal
+            }
+        })
+    });
```

Comment:
alternative suggestion:
after figuring out the cutoff, collect all items with weight< cutoff. Then filter for only weight == cutoff and do something like
```
let seeded_rng = rand::SeedableRng::seed_from_u64(block_height)
let (selected, _) = verifiers_with_cutoff_weight.partial_shuffle(seeded_rng, free_spot_count);
active_verifiers.extend_from_slice(selected)
```

Reply:
this way, you never have to sort and the shuffling should be faster than the hashing and comparison. This process is about fairness and not security (all verifiers are authorized), so relying on a seeded rng should be no concern